src/peer-store/README.md

```diff
@@ -1,3 +1,156 @@
 # Peerstore
 
-WIP
\ No newline at end of file
+Libp2p's Peerstore is responsible for keeping an updated register with the relevant information of the known peers. It should gather environment changes and be able to take decisions and notice interested parties of relevant changes. The Peerstore comprises four main components: `addressBook`, `keyBook`, `protocolBook` and `metadataBook`. These book components have similar characteristics with the `Javascript Map` implementation.
+
+The PeerStore needs to manage the high level operations on its inner books, have a job runner to trigger other books runners for data trimming or computations. Moreover, the peerStore should be responsible for noticing interested parties of relevant events, through its Event Emitter.
+
+## Peers Environment
+
+#### Sense
+
+Several libp2p subsystems will perform operations, which will gather relevant information about peers. Some operations might not have this as an end goal, but can also gather important data.
+
+In a libp2p node life, it will discover peers the existance of peers through its discovery protocols. In a typical discovery protocol, an address of the peer is discovered combined with its peer id. Once this happens, the `PeerStore` should collect this information for future (or immediate) usage by other subsystems. When the information is stored, the `PeerStore` should inform interested parties of the peer discovered (`peer` event).
+
+Taking into account a different scenario, a peer might perform/receive a dial request to/from a unkwown peer. In such a scenario, the `PeerStore` must store the peer's multiaddr once a connection is established. 
+
+(NOTE: this should be removed later)
+(currently we silently put it to the peerStore, without emitting events, as this logic exists in the `onConnected` callback from the upgrader. This way, we are never emitting the `peer` event when inbound connections happen, or a unkwown peer is dialed. Should we differentiate this?)
+
+After a connection is established with a peer, the Identify Service will act on this connection. A stream is created and peers exchange their information (listenMuldiaddrs and running protocols). Once this information is obtained, the PeerStore can collect the new data. In this specific case, we have a guarantee that this data is complete and updated, so the data stored in the PeerStore should be replaced (older and outdated data should disappear). However, if the recorded `multiaddrs` or `protocols` have changed, interested parties must be informed via `change:multiaddrs` or `change:protocols` events.
```

Comment:
Yes, I agree that we should consider the signed records spec + address origins (provenance). I will look at them deeply.

I thought that the identify should provide all the information (and updated), that's why I  think the replace should be the best option in this circumstance. This way, "If we've successfully dialed addresses of that peer we should probably avoid replacing them" if this multiaddr is not sent to us anymore, the peer should not be listening on it anymore...
Is there anything that I am not foreseeing?

Reply:
In theory, yes the peer should be advertising that address but there may be instances where that isn't happening. I think this will get better over time as we add support for announce/listen/noAnnounce addresses, but we don't support those at the moment so we shouldn't remove known good addresses unless they're actually bad. Keep in mind that when we deploy the updates, even if we have better support for that, older nodes won't have it yet so we risk removing good addresses. Addresses dialed via things like `ipfs swarm connect` come to mind as potential areas that we might end up removing good addresses for.